### PR TITLE
Implementation of the paper "Jointly Learning to Align and Translate with Transformer Models"

### PR DESCRIPTION
**Pull request by sarthakgarg**
This PR implements guided alignment training described in  "Jointly Learning to Align and Translate with Transformer Models (https://arxiv.org/abs/1909.02074)". 

In summary, it allows for training selected heads of the Transformer Model with external alignments computed by Statistical Alignment Toolkits. During inference, attention probabilities from the trained heads can be used to extract reliable alignments. In our work, we did not see any regressions in the translation performance because of guided alignment training.

### CODE REVIEW

**Comment by myleott** [comment]
Sorry for the delay in reviewing/merging this! This looks great!

A couple requests:
1) can you please add an integration test so that this functionality doesn't break in the future? You can look at the existing ones here: https://github.com/pytorch/fairseq/blob/master/tests/test_binaries.py#L36-L42. The general idea of these is to generate some dummy data, preprocess it, train the model for an epoch, and then do generation. It'd be great if you could add a new test there with the appropriate CLI args for the new alignment functionality.
2) can you please add a brief README in a new subfolder under `examples/`? At a minimum we'd just need some description of how to use the new args, as well as the bibtex for people to cite the paper. Ideally you could have a more detailed example (perhaps over the IWSLT De-En data for simplicity) showing how to preprocess, train, and generate with the new args, but this is optional. You could use the Scaling NMT one as a template if it's helpful: https://github.com/pytorch/fairseq/tree/master/examples/scaling_nmt
3) okay if we add a link to the paper in the main README?

File: fairseq/criterions/label_smoothed_cross_entropy_with_alignment.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+#
+# This source code is licensed under the license found in the LICENSE file in
+# the root directory of this source tree. An additional grant of patent rights
+# can be found in the PATENTS file in the same directory.
```
Comment (myleott) [inline fairseq/criterions/label_smoothed_cross_entropy_with_alignment.py]: nit: please update to the new header: https://github.com/pytorch/fairseq/blob/master/fairseq/criterions/label_smoothed_cross_entropy.py#L1-L4

Comment (sarthakgarg) [reply fairseq/criterions/label_smoothed_cross_entropy_with_alignment.py]: Done.

**Comment by sarthakgarg** [comment]
Thanks for the review!

1. I have added the integration test and a README under examples/joint_alignment_translation.
2. Adding the link to the main README would be great!

**Comment by sarthakgarg** [comment]
Hi Myle,
What is the status on merging this PR. Please let me know if any more changes need to be done.

Thanks!

File: examples/joint_alignment_translation/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+fairseq-train \
+    binarized \
+    --arch transformer_wmt_en_de_big_align --share-all-embeddings \
+    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 --activation-fn relu\
+    --lr 0.0002 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
+    --dropout 0.3 --attention-dropout 0.1 --weight-decay 0.0 \
+    --max-tokens 3500 --label-smoothing 0.1 \
+    --save-dir ./checkpoints --log-interval 1000 --max-update 60000 \
+    --keep-interval-updates -1 --save-interval-updates 0 \
+    --load-alignments --criterion label_smoothed_cross_entropy_with_alignment
```
Comment (stefan-it) [inline examples/joint_alignment_translation/README.md]: `\` is missing at the end of this line :)

Comment (sarthakgarg) [reply examples/joint_alignment_translation/README.md]: Fixed.

**Pull request closed without merging**